lib/vast/Dialect/HighLevel/Transforms/UDE.cpp

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) 2021-present, Trail of Bits, Inc.
+
+#include "vast/Dialect/HighLevel/Passes.hpp"
+
+VAST_RELAX_WARNINGS
+#include <mlir/Dialect/ControlFlow/IR/ControlFlowOps.h>
+#include <mlir/Dialect/Func/IR/FuncOps.h>
+
+#include <mlir/Rewrite/FrozenRewritePatternSet.h>
+#include <mlir/Transforms/DialectConversion.h>
+#include <mlir/Transforms/GreedyPatternRewriteDriver.h>
+VAST_UNRELAX_WARNINGS
+
+#include <gap/core/memoize.hpp>
+#include <gap/core/generator.hpp>
+
+#include <vast/Dialect/HighLevel/HighLevelUtils.hpp>
+#include <vast/Interfaces/AggregateTypeDefinitionInterface.hpp>
+
+#include <vast/Conversion/Common/Passes.hpp>
+#include <vast/Util/TypeUtils.hpp>
+
+#include "PassesDetails.hpp"
+
+#include <unordered_set>
+
+namespace vast::hl {
+
+#if !defined(NDEBUG)
+    constexpr bool debug_ude_pass = false;
+    #define VAST_UDE_DEBUG(...) VAST_REPORT_WITH_PREFIX_IF(debug_ude_pass, "[UDE] ", __VA_ARGS__)
+#else
+    #define VAST_UDE_DEBUG(...)
+#endif
+
+    constexpr auto keep_only_if_used = [] (auto, auto) { /* do not keep by default */ return false; };
+
```

Comment:
I don't understand the purpose of this lambda. Why is it preferred over simply returning `false` from the caller?